### PR TITLE
Feat/automatic induction slot page

### PR DESCRIPTION
**Pull request by nishant22029**
Issue - [137](https://github.com/coloredcow-admin/goonj-crm/issues/137)

Description  For Changes:
-  Created generate_induction_slots as the main function to determine available induction slots for a contact based on location, registration date, and existing induction activities.
      - If an existing induction activity for the contact is already Scheduled or Completed, the function now directly returns the activity’s status and date.
      - Incorporated location-specific logic to distinguish between physical and online inductions:
           - For specific cities in Bihar, Jharkhand, and Orissa, inductions default to a physical location.
           - Contacts in other cities or without specific office access receive online induction slots.
      - generate_slots Helper Function:

           - Added generate_slots as a helper function to dynamically create slots based on induction type, office information, and activity count. This function builds the slot data while ensuring that valid induction days and times 
- Added Changes in render.php file
     - Slot Generation: Initialized $source_contact_id from $action_target to generate induction slots personalized for each user.
     - Rendered the induction status and date/time if the slot is Scheduled or Completed.
     - Displayed available induction slots with day, date, and time. Each slot includes a dynamic booking link constructed with the following parameters:
source_contact_id: The ID of the contact initiating the booking.
slot_date: The selected date for the induction slot.
slot_time: The selected time for the induction slot.
induction_type: The type of induction (e.g., physical or online) associated with the slot.
     - Implemented logic to disable the booking button if the activity_count exceeds 20.
     - Enhanced the gb_goonj_blocks_custom_rewrite_rules function to include the induction-schedule rewrite rule.Refactored the existing rewrite rules into a loop, improving maintainability and reducing code duplication.

## Release Notes

- **New Features**
	- Introduced induction scheduling functionality, allowing users to view and book available induction slots.
	- Added support for online meeting links and detailed induction activity information.
	- New shortcode for handling induction slot details.
	- Enhanced integration with CiviCRM for improved contact and induction data retrieval.
	- Implemented a grid layout for displaying induction slots, optimized for various screen sizes.
	- Expanded token handling capabilities for induction details.
	- Added new CSS classes for improved styling of induction slots and action buttons.

- **Bug Fixes**
	- Improved validation and error handling for induction activities.

- **Documentation**
	- Updated theme metadata and CSS documentation for better usability.

- **Style**
	- Enhanced responsive design with new CSS classes for induction slot layouts.
	- Introduced new utility classes for improved styling across various screen sizes.

### CODE REVIEW

File: wp-content/civi-extensions/goonjcustom/goonjcustom.php [hunk 1]
```diff
@@ -126,6 +129,35 @@ function goonjcustom_evaluate_tokens(TokenValueEvent $e) {
       }
 
       $row->tokens('contact', 'inductionDetails', $inductionDetailsMarkup);
+			// Fetch induction activity details
+			$inductionActivity = \Civi\Api4\Activity::get(FALSE)
+					->addSelect('activity_date_time', 'Induction_Fields.Goonj_Office')
+					->addWhere('activity_type_id:name', '=', 'Induction')
+					->addWhere('source_contact_id', '=', $contactId)
+					->setLimit(1)
```
Comment (tarunnjoshi) [inline wp-content/civi-extensions/goonjcustom/goonjcustom.php:137]: Why are we using ->setLimit(1) here when we already have ->first()? Do we really need it, or should it be removed?

File: wp-content/civi-extensions/goonjcustom/goonjcustom.php [hunk 2]
```diff
@@ -126,6 +129,35 @@ function goonjcustom_evaluate_tokens(TokenValueEvent $e) {
...
+					->execute()
+					->first();
```
Comment (tarunnjoshi) [inline wp-content/civi-extensions/goonjcustom/goonjcustom.php:139]: Can we use ->single(); here insted of first?

File: wp-content/civi-extensions/goonjcustom/goonjcustom.php [hunk 3]
```diff
@@ -126,6 +129,35 @@ function goonjcustom_evaluate_tokens(TokenValueEvent $e) {
...
+			// Fetch office online meet link details if induction office is specified
+			$inductionOnlineMeetlink = '';
+			if ($inductionGoonjOffice) {
+					$officeDetails = \Civi\Api4\Contact::get(FALSE)
+							->addSelect('Goonj_Office_Details.Induction_Meeting_Access_Link')
+							->addWhere('contact_sub_type', 'CONTAINS', 'Goonj_Office')
+							->addWhere('id', '=', $inductionGoonjOffice)
+							->execute()
+							->first();
+
```
Comment (tarunnjoshi) [inline wp-content/civi-extensions/goonjcustom/goonjcustom.php:153]: Can we try to use single here?

File: wp-content/plugins/goonj-blocks/build/functions.php [hunk 4]
```diff
@@ -15,4 +15,207 @@ function gb_format_time_range($start_date, $end_date, $format = 'h:i A') {
...
+            ->execute();
+
+        // Return if no induction activity found
+        if ($inductionActivities->count() === 0) {
+            \Civi::log()->info('No activity found for contact', ['contactId' => $contactId]);
+            return [];
+        }
+
+        // If the induction activity status is 'Scheduled' or 'Completed', return the details
+        $inductionActivity = $inductionActivities->first();
```
Comment (tarunnjoshi) [inline wp-content/plugins/goonj-blocks/build/functions.php:59]: Can we use single on the API itself so we don't need to add first here

File: wp-content/plugins/goonj-blocks/build/functions.php [hunk 5]
```diff
@@ -15,4 +15,207 @@ function gb_format_time_range($start_date, $end_date, $format = 'h:i A') {
...
+        $contactStateId = intval($contactData['address_primary.state_province_id']);
+        $inductionSlotStartDate = (new DateTime($contactData['Individual_fields.Created_Date']))->modify('+1 day');
+        $physicalInductionType = 'Processing_Unit';
+        $onlineInductionType = 'Online_only_selected_by_Urban_P';
+
+        // List of states with both physical and online inductions based on cities
+        $statesWithMixedInductionTypes = \Civi\Api4\StateProvince::get(FALSE)
+            ->addWhere('country_id.name', '=', 'India')
+            ->addWhere('name', 'IN', ['Bihar', 'Jharkhand', 'Orissa'])
+            ->setLimit(3)
```
Comment (tarunnjoshi) [inline wp-content/plugins/goonj-blocks/build/functions.php:76]: why we are set limit 3 here we only want 3 data?

File: wp-content/plugins/goonj-blocks/build/functions.php [hunk 6]
```diff
@@ -15,4 +15,209 @@ function gb_format_time_range($start_date, $end_date, $format = 'h:i A') {
...
+        // Check for mixed induction type states and generate slots accordingly
+        if (in_array($contactStateId, $statesWithMixedInductionTypes)) {
+            $contactCity = isset($contact['address_primary.city']) ? strtolower($contact['address_primary.city']) : '';
+            if (in_array($contactCity, ['patna', 'ranchi', 'bhubaneshwar'])) {
+                // generate physical slot for these cities
+                return generate_slots($assignedOfficeId, 30, $physicalInductionType, $inductionSlotStartDate);
+            }
+            // generate online induction slots for other cities
+            return generate_slots($assignedOfficeId, 30, $onlineInductionType, $inductionSlotStartDate);
+        }
```
[the remark this answers is not in the document]
Comment (tarunnjoshi) [reply wp-content/plugins/goonj-blocks/build/functions.php:91]: Can we add this

File: wp-content/plugins/goonj-blocks/build/functions.php [hunk 7]
```diff
@@ -15,4 +15,207 @@ function gb_format_time_range($start_date, $end_date, $format = 'h:i A') {
...
+        // Generate slots
+        generateActivitySlots($slots, $maxSlots, $validInductionDays, $hour, $minute, $startDate, $scheduledActivityDates, $slotCount, $highActivityCountDays, $inductionType);
+
+        if ($highActivityCountDays >= 23) {
+            $slotCount = 0;
+            $startDate = new DateTime(end($slots)['date']);
+            generateActivitySlots($slots, $highActivityCountDays, $validInductionDays, $hour, $minute, $startDate, $scheduledActivityDates, $slotCount, $highActivityCountDays, $inductionType);
+        }
+
+        \Civi::log()->info('Activity slots created', ['slots' => $slots]);
```
Comment (tarunnjoshi) [inline wp-content/plugins/goonj-blocks/build/functions.php:179]: Should we need this here?

File: wp-content/plugins/goonj-blocks/build/functions.php [hunk 8]
```diff
@@ -15,4 +15,209 @@ function gb_format_time_range($start_date, $end_date, $format = 'h:i A') {
...
+            'officeId' => $assignedOfficeId,
+            'error' => $e->getMessage()
+        ]);
+        return [];
+    }
+}
+
+function generateActivitySlots(&$slots, $maxSlots, $validInductionDays, $hour, $minute, $startDate, $scheduledActivityDates, &$slotCount, &$highActivityCountDays) {
+    for ($i = 0; $slotCount < $maxSlots; $i++) {
+        $date = (clone $startDate)->modify("+{$i} days");
```
[the remark this answers is not in the document]
Comment (tarunnjoshi) [reply wp-content/plugins/goonj-blocks/build/functions.php:192]: can you please check this?

**Review by tarunnjoshi — CHANGES_REQUESTED** [review]

**Comment by tarunnjoshi** [comment]
This PR contains a lot of code, making it time-consuming to review. Can we try to split it into separate PRs? it's easier for the reviewer to review it.

File: wp-content/plugins/goonj-blocks/build/functions.php (hunk 6 above)
[the remark this answers is not in the document]
Comment (nishant22029) [reply wp-content/plugins/goonj-blocks/build/functions.php:91]: changes done

File: wp-content/plugins/goonj-blocks/build/functions.php (hunk 8 above)
[the remark this answers is not in the document]
Comment (nishant22029) [reply wp-content/plugins/goonj-blocks/build/functions.php:192]: sure done changes

File: wp-content/plugins/goonj-blocks/src/functions.php [hunk 9]
```diff
@@ -15,4 +15,207 @@ function gb_format_time_range($start_date, $end_date, $format = 'h:i A') {
...
+        // Check for mixed induction type states and generate slots accordingly
+        if (in_array($contactStateId, $statesWithMixedInductionTypes)) {
+            $contactCity = isset($contact['address_primary.city']) ? strtolower($contact['address_primary.city']) : '';
+            if (in_array($contactCity, ['patna', 'ranchi', 'bhubaneshwar'])) {
+                // generate physical slot for these cities
+                return generate_slots($assignedOfficeId, 30, $physicalInductionType, $inductionSlotStartDate);
+            }
+            // generate online induction slots for other cities
+            return generate_slots($assignedOfficeId, 30, $onlineInductionType, $inductionSlotStartDate);
+        }
```
[the remark this answers is not in the document]
Comment (nishant22029) [reply wp-content/plugins/goonj-blocks/src/functions.php:91]: changes done

File: wp-content/plugins/goonj-blocks/build/functions.php (hunk 4 above)
Comment (nishant22029) [reply wp-content/plugins/goonj-blocks/build/functions.php:59]: sure

File: wp-content/plugins/goonj-blocks/build/functions.php (hunk 5 above)
Comment (nishant22029) [reply wp-content/plugins/goonj-blocks/build/functions.php:76]: yes

File: wp-content/plugins/goonj-blocks/build/functions.php [hunk 10]
```diff
@@ -15,4 +15,208 @@ function gb_format_time_range($start_date, $end_date, $format = 'h:i A') {
...
+        $contactStateId = intval($contactData['address_primary.state_province_id']);
+        $inductionSlotStartDate = (new DateTime($contactData['Individual_fields.Created_Date']))->modify('+1 day');
+        $physicalInductionType = 'Processing_Unit';
+        $onlineInductionType = 'Online_only_selected_by_Urban_P';
+        $defaultMaxSlot = 30;
+
+        // List of states with both physical and online inductions based on cities
+        $statesWithMixedInductionTypes = \Civi\Api4\StateProvince::get(FALSE)
+            ->addWhere('country_id.name', '=', 'India')
+            ->addWhere('name', 'IN', ['Bihar', 'Jharkhand', 'Orissa'])
```
Comment (tarunnjoshi) [inline wp-content/plugins/goonj-blocks/build/functions.php:76]: If we are setting the addWhere here and fetching 3 states, should we need to add this line `->setLimit(3)`, can you please remove ->setLimit(3) it and then test it working fine or not

File: wp-content/plugins/goonj-blocks/build/functions.php [hunk 11]
```diff
@@ -15,4 +15,208 @@ function gb_format_time_range($start_date, $end_date, $format = 'h:i A') {
...
+    $slotCount = 0;
+    $highActivityCountDays = 0;
+
+    try {
+        // Fetch office induction details for induction scheduling
+        $officeDetails = \Civi\Api4\Contact::get(FALSE)
+            ->addSelect('display_name', 'Goonj_Office_Details.Physical_Induction_Slot_Days:name', 'Goonj_Office_Details.Physical_Induction_Slot_Time', 'Goonj_Office_Details.Online_Induction_Slot_Days:name', 'Goonj_Office_Details.Online_Induction_Slot_Time')
+            ->addWhere('contact_sub_type', 'CONTAINS', 'Goonj_Office')
+            ->addWhere('id', '=', $assignedOfficeId)
+            ->execute()->first();
```
Comment (tarunnjoshi) [inline wp-content/plugins/goonj-blocks/build/functions.php:137]: Can you please check everwhere and change it to single whenever needed

File: wp-content/plugins/goonj-blocks/build/functions.php [hunk 12]
```diff
@@ -15,4 +15,208 @@ function gb_format_time_range($start_date, $end_date, $format = 'h:i A') {
...
+        list($hour, $minute) = explode(':', $inductionTime);
+
+        // Fetch already scheduled induction activities at the assigned Goonj office
+        $scheduledActivities = \Civi\Api4\Activity::get(FALSE)
+            ->addSelect('activity_date_time')
+            ->addWhere('activity_type_id:name', '=', 'Induction')
+            ->addWhere('status_id:name', '=', 'Scheduled')
+            ->addWhere('Induction_Fields.Goonj_Office', '=', $assignedOfficeId)
+            ->addWhere('activity_date_time', '>', (new DateTime('today midnight'))->format('Y-m-d H:i:s'))
+            ->setLimit(60)
```
Comment (tarunnjoshi) [inline wp-content/plugins/goonj-blocks/build/functions.php:162]: Why we needed this, are we fetching only 60 scheduled induction activities at the assigned Goonj office?

File: wp-content/plugins/goonj-blocks/goonj-blocks.php [hunk 13]
```diff
@@ -95,6 +89,18 @@ function gb_goonj_blocks_check_action_target_exists() {
 	);
 
 	switch ( $target ) {
+		case 'induction-schedule':
+			$result = \Civi\Api4\Contact::get(FALSE)
+				->addWhere('id', '=', $id)
+				->setLimit(1)
```
Comment (tarunnjoshi) [inline wp-content/plugins/goonj-blocks/goonj-blocks.php:95]: Needed?

File: wp-content/plugins/goonj-blocks/src/functions.php [hunk 14]
```diff
@@ -15,4 +15,209 @@ function gb_format_time_range($start_date, $end_date, $format = 'h:i A') {
...
+        $inductionSlotStartDate = (new DateTime($contactData['Individual_fields.Created_Date']))->modify('+1 day');
+        $physicalInductionType = 'Processing_Unit';
+        $onlineInductionType = 'Online_only_selected_by_Urban_P';
+        $defaultMaxSlot = 30;
+
+        // List of states with both physical and online inductions based on cities
+        $statesWithMixedInductionTypes = \Civi\Api4\StateProvince::get(FALSE)
+            ->addWhere('country_id.name', '=', 'India')
+            ->addWhere('name', 'IN', ['Bihar', 'Jharkhand', 'Orissa'])
+            ->setLimit(3)
```
Comment (tarunnjoshi) [inline wp-content/plugins/goonj-blocks/src/functions.php:77]: same feedback as above

File: wp-content/plugins/goonj-blocks/src/functions.php [hunk 15]
```diff
@@ -15,4 +15,209 @@ function gb_format_time_range($start_date, $end_date, $format = 'h:i A') {
...
+        list($hour, $minute) = explode(':', $inductionTime);
+
+        // Fetch already scheduled induction activities at the assigned Goonj office
+        $scheduledActivities = \Civi\Api4\Activity::get(FALSE)
+            ->addSelect('activity_date_time')
+            ->addWhere('activity_type_id:name', '=', 'Induction')
+            ->addWhere('status_id:name', '=', 'Scheduled')
+            ->addWhere('Induction_Fields.Goonj_Office', '=', $assignedOfficeId)
+            ->addWhere('activity_date_time', '>', (new DateTime('today midnight'))->format('Y-m-d H:i:s'))
+            ->setLimit(60)
```
Comment (tarunnjoshi) [inline wp-content/plugins/goonj-blocks/src/functions.php:162]: can you please let me know the purpose here

File: wp-content/plugins/goonj-blocks/src/functions.php [hunk 16]
```diff
@@ -15,4 +15,209 @@ function gb_format_time_range($start_date, $end_date, $format = 'h:i A') {
...
+        // Generate slots
+        generateActivitySlots($slots, $maxSlots, $validInductionDays, $hour, $minute, $startDate, $scheduledActivityDates, $slotCount, $highActivityCountDays, $inductionType);
+
+        if ($highActivityCountDays >= 23) {
+            $slotCount = 0;
+            $startDate = new DateTime(end($slots)['date']);
+            generateActivitySlots($slots, $highActivityCountDays, $validInductionDays, $hour, $minute, $startDate, $scheduledActivityDates, $slotCount, $highActivityCountDays, $inductionType);
+        }
+
+        \Civi::log()->info('Activity slots created', ['slots' => $slots]);
```
Comment (tarunnjoshi) [inline wp-content/plugins/goonj-blocks/src/functions.php:180]: Are we needed this?

File: wp-content/plugins/goonj-blocks/src/functions.php [hunk 17]
```diff
@@ -15,4 +15,209 @@ function gb_format_time_range($start_date, $end_date, $format = 'h:i A') {
...
+            if ($activityCount > 20) {
+                $highActivityCountDays++;
+            }
+        }
+    }
+}
+
+
+
+
```
Comment (tarunnjoshi) [inline wp-content/plugins/goonj-blocks/src/functions.php:223]: Remove extra lines

File: wp-content/themes/goonj-crm/engine/shortcodes.php [hunk 18]
```diff
@@ -93,7 +92,7 @@ function goonj_pu_activity_button() {
 
 	try {
 		$activity = \Civi\Api4\Activity::get(false)
-			->addSelect('source_contact_id', 'Office_Visit.Goonj_Processing_Center', 'Material_Contribution.Goonj_Office', 'activity_type_id:name')
+			->addSelect('custom.*', 'source_contact_id', 'Office_Visit.Goonj_Processing_Center', 'Material_Contribution.Goonj_Office', 'activity_type_id:name')
```
Comment (tarunnjoshi) [inline wp-content/themes/goonj-crm/engine/shortcodes.php:95]: are we need this change?

**Review by tarunnjoshi — CHANGES_REQUESTED** [review]

File: wp-content/plugins/goonj-blocks/src/functions.php (hunk 16 above)
Comment (nishant22029) [reply wp-content/plugins/goonj-blocks/src/functions.php:180]: removed

File: wp-content/plugins/goonj-blocks/goonj-blocks.php (hunk 13 above)
Comment (nishant22029) [reply wp-content/plugins/goonj-blocks/goonj-blocks.php:95]: yes to get one record only

File: wp-content/plugins/goonj-blocks/build/functions.php (hunk 12 above)
Comment (nishant22029) [reply wp-content/plugins/goonj-blocks/build/functions.php:162]: removed this

File: wp-content/plugins/goonj-blocks/build/functions.php (hunk 11 above)
Comment (nishant22029) [reply wp-content/plugins/goonj-blocks/build/functions.php:137]: sure

File: wp-content/plugins/goonj-blocks/build/functions.php (hunk 10 above)
Comment (nishant22029) [reply wp-content/plugins/goonj-blocks/build/functions.php:76]: removed

File: wp-content/plugins/goonj-blocks/build/functions.php [hunk 19]
```diff
@@ -15,4 +15,203 @@ function gb_format_time_range($start_date, $end_date, $format = 'h:i A') {
...
+ * Generate induction slots based on contact's location, registration date, and induction activity details.
+ * 
+ * @param int|null $contactId The ID of the contact for whom to generate slots.
+ * @param int $days The number of days to look ahead for slot generation (default is 30 days).
+ * @return array An array of induction slots or empty if criteria aren't met.
+ */
+function generate_induction_slots($contactId = null, $days = 30) {
+    // Validate contactId
+    if (empty($contactId) || !is_numeric($contactId)) {
+        \Civi::log()->warning('Invalid contactId', ['contactId' => $contactId]);
```
Comment (tarunnjoshi) [inline wp-content/plugins/goonj-blocks/build/functions.php:29]: As we are using a warning here are we showing the message to the user? If not then why we are not using Debug

File: wp-content/plugins/goonj-blocks/build/functions.php [hunk 20]
```diff
@@ -15,4 +15,203 @@ function gb_format_time_range($start_date, $end_date, $format = 'h:i A') {
...
+
+    try {
+        // Fetch necessary contact details for slot generation
+        $contactData = \Civi\Api4\Contact::get(FALSE)
+            ->addSelect('address_primary.state_province_id', 'address_primary.city', 'Individual_fields.Created_Date')
+            ->addWhere('id', '=', $contactId)
+            ->execute()->single();
+
+        if (empty($contactData)) {
+            \Civi::log()->warning('Contact not found', ['contactId' => $contactId]);
```
Comment (tarunnjoshi) [inline wp-content/plugins/goonj-blocks/build/functions.php:41]: same here

File: wp-content/plugins/goonj-blocks/build/functions.php [hunk 21]
```diff
@@ -15,4 +15,203 @@ function gb_format_time_range($start_date, $end_date, $format = 'h:i A') {
...
+    $highActivityCountDays = 0;
+
+    try {
+        // Fetch office induction details for induction scheduling
+        $officeDetails = \Civi\Api4\Contact::get(FALSE)
+            ->addSelect('display_name', 'Goonj_Office_Details.Physical_Induction_Slot_Days:name', 'Goonj_Office_Details.Physical_Induction_Slot_Time', 'Goonj_Office_Details.Online_Induction_Slot_Days:name', 'Goonj_Office_Details.Online_Induction_Slot_Time')
+            ->addWhere('contact_sub_type', 'CONTAINS', 'Goonj_Office')
+            ->addWhere('id', '=', $assignedOfficeId)
+            ->execute()
+            ->single();
```
Comment (tarunnjoshi) [inline wp-content/plugins/goonj-blocks/build/functions.php:137]:
```suggestion
            ->execute()->single();
```

File: wp-content/plugins/goonj-blocks/src/functions.php [hunk 22]
```diff
@@ -15,4 +15,203 @@ function gb_format_time_range($start_date, $end_date, $format = 'h:i A') {
...
+    $highActivityCountDays = 0;
+
+    try {
+        // Fetch office induction details for induction scheduling
+        $officeDetails = \Civi\Api4\Contact::get(FALSE)
+            ->addSelect('display_name', 'Goonj_Office_Details.Physical_Induction_Slot_Days:name', 'Goonj_Office_Details.Physical_Induction_Slot_Time', 'Goonj_Office_Details.Online_Induction_Slot_Days:name', 'Goonj_Office_Details.Online_Induction_Slot_Time')
+            ->addWhere('contact_sub_type', 'CONTAINS', 'Goonj_Office')
+            ->addWhere('id', '=', $assignedOfficeId)
+            ->execute()
+            ->single();
```
Comment (tarunnjoshi) [inline wp-content/plugins/goonj-blocks/src/functions.php:137]:
```suggestion
            ->execute()->single();
```

**Review by tarunnjoshi — APPROVED** [review]

File: wp-content/plugins/goonj-blocks/src/functions.php (hunk 15 above)
Comment (tarunnjoshi) [reply wp-content/plugins/goonj-blocks/src/functions.php]: @nishant22029 ?

File: wp-content/plugins/goonj-blocks/build/functions.php (hunk 19 above)
Comment (nishant22029) [reply wp-content/plugins/goonj-blocks/build/functions.php:29]: updated and used info()

**Pull request merged**